### PR TITLE
Form content

### PR DESCRIPTION
**Pull request by swwol**

### CODE REVIEW

File: Sources/Webservice/HTTP/Headers/ContentType.swift [hunk 1]
```diff
@@ -14,20 +14,45 @@ public protocol ContentTypeDecoder {
     func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T
 }
 
+public enum CharSet {
```
Comment (timsearle) [inline Sources/Webservice/HTTP/Headers/ContentType.swift]: Does this make sense as a top-level type? Seems pretty linked to the JSON content-type? Could nest

File: Sources/Webservice/HTTP/Headers/ContentType.swift [hunk 2]
```diff
@@ -14,20 +14,45 @@ public protocol ContentTypeDecoder {
...
+    case utf8
+    case iso88591
+    case custom(String)
+
+    var stringValue: String {
+        switch self {
+        case .utf8:
+            return "\"UTF-8\""
```
Comment (timsearle) [inline Sources/Webservice/HTTP/Headers/ContentType.swift]: How come these are all in escaped quotes?

File: Sources/Webservice/HTTP/Headers/ContentType.swift [hunk 3]
```diff
@@ -14,20 +14,45 @@ public protocol ContentTypeDecoder {
...
         return JSONEncoder()
     }
 
-    public var decoder: ContentTypeDecoder {
+    public var decoder: ContentTypeDecoder? {
         return JSONDecoder()
     }
 
-    public init() {}
+    public init( charSet: CharSet? = nil ) {
```
Comment (timsearle) [inline Sources/Webservice/HTTP/Headers/ContentType.swift]: Strange spacing

File: Sources/Resource/FormEncoder.swift [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+        let keys = codingKey.map { $0.stringValue }
+        let bracketedKeys = keys.dropFirst().map { "[\($0)]" }.joined()
+        let key = keys[0].appending(bracketedKeys)
+        strings[key] = value
+    }
+}
+
+extension String {
+    func escaped() -> String {
+        let plussesReplaced = self.replacingOccurrences(of: "+", with: "%2B")
```
Comment (timsearle) [inline Sources/Resource/FormEncoder.swift]: Why not use the Foundation framework methods to handle percent encoding/decoding? Or use URLComponents?

File: Sources/Resource/FormEncoder.swift [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+        let ampersandsReplaced = plussesReplaced.replacingOccurrences(of: "&", with: "%26")
+        let spacesReplaced = ampersandsReplaced.replacingOccurrences(of: " ", with: "+")
+        return spacesReplaced.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)!
+    }
+
+    func unescaped() -> String {
+     let putBackSpaces = self.replacingOccurrences(of: "+", with: " ")
```
Comment (timsearle) [inline Sources/Resource/FormEncoder.swift]: Inconsistent spacing here

File: Sources/Resource/FormEncoder.swift [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+}
+
+final class FormData {
+    private(set) var strings: [String: String] = [:]
+
+    func encode(key codingKey: [CodingKey], value: String) throws {
+        guard !codingKey.isEmpty else { throw FormEncoder.Error.missingKeyForValue }
+        let keys = codingKey.map { $0.stringValue }
+        let bracketedKeys = keys.dropFirst().map { "[\($0)]" }.joined()
+        let key = keys[0].appending(bracketedKeys)
```
Comment (timsearle) [inline Sources/Resource/FormEncoder.swift]: This subscript looks unsafe

File: Sources/Resource/FormEncoder.swift [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+        return container
+    }
+
+    mutating func superEncoder() -> Encoder {
+        let formEncoding = FormEncoding(to: data)
+        return formEncoding
+    }
+}
+
+private struct FormSingleValueEncoding: SingleValueEncodingContainer {
```
Comment (timsearle) [inline Sources/Resource/FormEncoder.swift]: Any chance these could be split into different files seeing as they don't offer any value currently? Also would reduce the line length of FormEncoder.swift

File: NetworkingTests/Webservice/FormEncoderTests.swift [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        let watch = Product(name: "Apple Watch", price: 500, info: "Series 4")
+
+        let products =  [ iPhone, macBook, watch]
+
+        let formEncoder = FormEncoder()
+        do {
+            let formString = try formEncoder.stringEncode(products)
+            print(formString)
+        } catch FormEncoder.Error.unsupportedDataFormat {
+            print("test passed")
```
Comment (timsearle) [inline NetworkingTests/Webservice/FormEncoderTests.swift]: There is an XCTAssertThrows, should be using that if validating errors are thrown

File: NetworkingTests/Webservice/FormEncoderTests.swift [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+        }
+    }
+
+    func testSingleUInt64Encoding() {
+        let formEncoder = FormEncoder()
+        do {
+            let formString = try formEncoder.stringEncode(UInt64(1))
+            print(formString)
+        } catch FormEncoder.Error.unsupportedDataFormat {
+            print("test passed")
```
Comment (timsearle) [inline NetworkingTests/Webservice/FormEncoderTests.swift]: XCTAssertThrows

**Review by timsearle — CHANGES_REQUESTED** [review]
Looks great! Glad we got the encoding working. Please improve the test coverage so it hasn't drop so much and there are a few comments/questions throughout.

File: Sources/Resource/FormEncoder.swift (hunk 4 above)
Comment (swwol) [reply Sources/Resource/FormEncoder.swift]: I am using the addingPercentEncoding method at the end, but first replacing spaces with plusses, ss that is how they are represented in form data, rather that as %20.
Also + and & are excluded from the .urlQueryAllowed characterset, so I am manually encoding them. (After first swapping spaces for +s)

File: Sources/Webservice/HTTP/Headers/ContentType.swift (hunk 2 above)
Comment (swwol) [reply Sources/Webservice/HTTP/Headers/ContentType.swift]: thats how it is defined here..
https://www.w3schools.com/tags/att_meta_charset.asp

File: Sources/Resource/FormEncoder/FormData.swift [hunk 10]
```diff
@@ -0,0 +1,3 @@
+//// Created by Nice Agency
```
Comment (timsearle) [inline Sources/Resource/FormEncoder/FormData.swift]: Could you delete all of these comment headers from each file?

**Comment by timsearle** [comment]
Some project structure and naming comments now that I've figured out how to clone down the PR from GitHub:

Rename FormContentType -> FormURLEncodedContentType (this could also be moved under the Webservice > HTTP  > Headers group.

Rename FormEncoder -> FormURLEncoder, including the group name

String+Extension.swift can be moved to where the other extensions live. The name should be formURLEncoded().

Basically instead of Form it should be FormURL everywhere to be explicit as to what these types do.

The FormEncoderTests.swift seem to be in the Webservice group on the test target, this doesn't align with where they are in the code structure. They should be under resource and perhaps create a new group. Also can be renamed appropriately.

File: NetworkingTests/Webservice/FormEncoderTests.swift [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        let macBook = Product(name: "Mac Book Pro", price: 2_000, info: "Early 2019")
+        let watch = Product(name: "Apple Watch", price: 500, info: "Series 4")
+
+        let products =  ["phone": iPhone, "mac": macBook, "watch": watch]
+
+        let formEncoder = FormEncoder()
+        do {
+            //swiftlint:disable line_length
+            let formString = try formEncoder.stringEncode(products)
+            print (formString)
```
Comment (timsearle) [inline NetworkingTests/Webservice/FormEncoderTests.swift]: Can remove all `print()` calls.

File: NetworkingTests/Webservice/FormEncoderTests.swift [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    func testNestedEncoding() {
+        let iPhone = Product(name: "iPhone X", price: 1_000, info: "Our best iPhone yet!")
+        let macBook = Product(name: "Mac Book Pro", price: 2_000, info: "Early 2019")
+        let watch = Product(name: "Apple Watch", price: 500, info: "Series 4")
+
+        let products =  ["phone": iPhone, "mac": macBook, "watch": watch]
+
+        let formEncoder = FormEncoder()
+        do {
+            //swiftlint:disable line_length
```
Comment (timsearle) [inline NetworkingTests/Webservice/FormEncoderTests.swift]: This will disable line_length for the whole file. You could put the test string into a multi-line string literal

File: NetworkingTests/Webservice/FormEncoderTests.swift [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+            let formString = try formEncoder.stringEncode(products)
+            print (formString)
+             XCTAssertEqual(formString, "mac%5Binfo%5D=Early+2019&mac%5Bname%5D=Mac+Book+Pro&mac%5Bprice%5D=2000.0&phone%5Binfo%5D=Our+best+iPhone+yet!&phone%5Bname%5D=iPhone+X&phone%5Bprice%5D=1000.0&watch%5Binfo%5D=Series+4&watch%5Bname%5D=Apple+Watch&watch%5Bprice%5D=500.0")
+        } catch {
+            print("Encoding failed: \(error)")
```
Comment (timsearle) [inline NetworkingTests/Webservice/FormEncoderTests.swift]: Rather than do, try, catch. You can make the test function `throws`, and XCTest will catch the error and treat it as a test failure. This should clean up a lot of lines of code from these tests

**Review by timsearle — APPROVED** [review]
